qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -186,10 +192,55 @@ def result(self, timeout=None, wait=5):
         except ApiError as api_err:
             raise JobError(str(api_err))
 
-        if self._is_device and self.status() == JobStatus.DONE:
-            _reorder_bits(job_data)
-
         # Build the Result.
+        experiment_results = []
+        if 'result' in job_data:  # qobj job
```

Comment:
I like the idea of recognizing both types of results during the transition between formats since *we can not ensure we will be receiving the new Result format if, and only if sending the new Qobj*.

If we embrace the idea of having different classes per each version of the job, we could add a class method `result_from_api_response(cls, ...)` to build the proper `Result` instance from the API response and use a function to decide when to use each of the methods:

 ```python
return _result_from_api_response(job_data, self._id, self._backend_name, self._is_device)
```

And outside the functions:

```python
def _result_from_api_response(response, id, backend_name, is_device):
  if 'result' in response:
    return IBMQjob.result_from_api_response(response)

  if 'qasms' in response:
    if is_device:
      _reorder_bits(response)
    return DeprecatedIBMQJob.result_from_api_response(response, id, backend_name)
  
  raise JobError('Unrecognized result type...')
```

Each `result_from_api_response()` implementation will encapsulate the behaviour for parsing the API response according to the kind of result it is expecting.

Reply:
sounds good